src/Lean/Data/Lsp/Basic.lean

```diff
@@ -78,6 +78,18 @@ structure Command where
   arguments? : Option (Array Json) := none
   deriving ToJson, FromJson
 
+/-- A snippet string is a template which allows to insert text
+and to control the editor cursor when insertion happens.
```

Comment:
Style nitpick: "allows" should always take a direct object.


```suggestion
/-- A snippet string is a template that simultaneously provides a string to be inserted
and a set of positions for the editor cursor within the inserted string.
```

Reply:
Fwiw this is copy-pasta from [VSC docs](https://code.visualstudio.com/api/references/vscode-api#SnippetString), but why not improve on their wording :)

Reply:
Actually, how about just replacing the entire docstring with a link to that? It will not fix the wording, but it will update automatically whenever they happen to change the docs; additionally, I don't entirely understand the details of their syntax, so am not comfortable trying to explain its minutiae.

Reply:
I did a bit more reading, and it seems that the syntax (and the "tab stop" terminology) comes [from TextMate](https://macromates.com/manual/en/snippets) - given that it's established terminology, I think we should just use it, and attempt to resolve reader confusion directly. I find the TextMate docs to be easier to understand than the ones you linked to, and the [non-API VSC docs](https://code.visualstudio.com/docs/editor/userdefinedsnippets) are even better. But neither of these describe [LSP snippets](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax), which are (as far as I can tell) as subset of both TextMate and VSC snippets.

Anyway, with all that background, I'd actually propose a different modification here that acknowledges the overloaded terminology but doesn't try to avoid it, and that calls users' attention to the specific syntax that LSP supports. In particular, the variables supported in LSP are fewer than those in VS Code, LSP regexes have fewer features than Javascript regexes,

What about this alternative approach?
```
A snippet is a template that includes instructions for subsequent edits once the string is inserted, expressed in a domain-specific language based on one from TextMate. Snippets may include:

 * Designated positions for subsequent user input, called "tab stops" after their most frequently-used keybinding, are denoted with `$1`, `$2`, and so forth. `$0` denotes the final cursor position after all edits are completed, defaulting to the end of the string. Snippet tab stops are unrelated to tab stops used for indentation.
 * Tab stops with default values, called _placeholders_, written `${1:default}`. The default may itself contain a tab stop or a further placeholder, and multiple options to select from may be provided by surrounding them with `|`s and separating them with `,`, as in `${1|if $2 then $3 else $4,if let $2 := $3 then $4 else $5|}`. 
 * One of a set of predefined variables that are replaced with values such as the current line number (`$TM_LINE_NUMBER`) or the text that was selected when the snippet was invoked (`$TM_SELECTED_TEXT`).
 * Formatting instructions to modify variables using regular expressions or a set of predefined filters.

The full syntax and semantics of snippets, including the available variables and the rules for escaping control characters, are described in the [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax).
```

Reply:
Done (with some rewording)!